Gem/Code/Source/GameLift/GameLiftServerComponent.cpp

```diff
@@ -0,0 +1,211 @@
+        return true;
+    }
+
+    bool GameLiftServerSystemComponent::OnCreateSessionBegin(const AzFramework::SessionConfig& sessionConfig)
+    {
+        m_maxPlayer = sessionConfig.m_maxPlayer;
+
+        return true;
+    }
+
+    void GameLiftServerSystemComponent::OnCreateSessionEnd()
+    {
+        StartBackfillProcess();
+    }
+
+    bool GameLiftServerSystemComponent::OnDestroySessionBegin()
+    {
+        StopBackfillProcess();
+
+        return true;
+    }
+
+    void GameLiftServerSystemComponent::OnDestroySessionEnd()
+    {
+        m_maxPlayer = 0;
+    }
+
+    void GameLiftServerSystemComponent::OnUpdateSessionBegin(
+        const AzFramework::SessionConfig& sessionConfig, const AZStd::string& updateReason)
+    {
+        AZ_UNUSED(sessionConfig);
+        if (updateReason == "BACKFILL_FAILED" || updateReason == "BACKFILL_TIMED_OUT" || updateReason == "BACKFILL_CANCELLED")
+        {
+            m_backfillComplete = true;
+            m_waitEvent.release();
+        }
+    }
+
+    void GameLiftServerSystemComponent::OnUpdateSessionEnd()
+    {
+        AZ_TracePrintf("MultiplayerSample", "Multiplayer sample server manual backfill has completed");
+
+        m_backfillComplete = true;
+    }
+
+    void GameLiftServerSystemComponent::StartBackfillProcess()
+    {
+        bool manualBackfill = false;
+        const auto console = AZ::Interface<AZ::IConsole>::Get();
+        if (console->GetCvarValue("sv_manualBackfill", manualBackfill) != AZ::GetValueResult::Success)
+        {
+            AZ_Error("GameLiftClientSystemComponent", false, "Failed to get sv_manualBackfill");
+            return;
+        }
+
+        if (!manualBackfill || !m_backfillProcessTerminated)
+        {
+            return;
+        }
+
+        AZ_TracePrintf("MultiplayerSample", "Start multiplayer sample server manual backfill");
+        m_backfillProcessTerminated = false;
+        m_backfillThread = AZStd::thread(AZStd::bind(&GameLiftServerSystemComponent::BackfillProcess, this));
+    }
+
+    void GameLiftServerSystemComponent::StopBackfillProcess()
+    {
+        bool manualBackfill = false;
+        const auto console = AZ::Interface<AZ::IConsole>::Get();
+        if (console->GetCvarValue("sv_manualBackfill", manualBackfill) != AZ::GetValueResult::Success)
+        {
+            AZ_Error("GameLiftClientSystemComponent", false, "Failed to get sv_manualBackfill");
+            return;
+        }
+
+        if (!manualBackfill || m_backfillProcessTerminated)
+        {
+            return;
+        }
+
+        m_backfillProcessTerminated = true;
+        m_waitEvent.release();
+        if (m_backfillThread.joinable())
+        {
+            m_backfillThread.join();
+        }
+    }
+
+    void GameLiftServerSystemComponent::BackfillProcess()
+    {
+        while (!m_backfillProcessTerminated)
+        {
+            m_waitEvent.try_acquire_for(AZStd::chrono::seconds(30));
+
+            if (!m_backfillComplete || !HasEmptySlots())
```

Comment:
So the idea is to poll and check for empty slots on a side thread? There are events on the IMultiplayer interface you can subscribe to that let you react to slot availability vs polling. If those aren't quite sufficient, IMultiplayer should be updated such that they are

Reply:
Yeah, that's the idea here. GameLift suggests to initiate a backfill request whenever a matched game has one or more empty player slots. I'll check the IMultiplayer interface on it.